### DISCUSSION TITLE
SO_REUSEPORT

### DISCUSSION THREAD

**Discussion by Qarasique** [0xERR0R/blocky#554] [Q&A]
Does the blocky have SO_REUSEPORT support ? (for more performance)

**Comment by 0xERR0R** [comment]
The underlying dns library has support for port reusage, it this case SO_REUSEPORT will be set. This feature is currently is not used in blocky. I'll try it out and will do some performance checks

**Reply by ThinkChaos** [reply]
I don't think SO_REUSEPORT has a direct perfomance impact.  
Maybe @Qarasique wants to run two instances? But I don't see why that'd improve performance either.

**Reply by 0xERR0R** [reply]
There is some interesting information https://github.com/miekg/dns/issues/654 (and also linked pr). I think it could improve performance on high load

**Reply by ThinkChaos** [reply]
Yeah the main thing there is increasing concurrency with more goroutines and parallelism with more cores, and indeed SO_REUSEPORT allows that easily.  
I meant `SO_REUSEPORT`, the socket option, itself doesn't change anything.

I'm all for having a setting to set the number of goroutines, with a dynamic default.

**Reply by Qarasique** [reply]
I thought that we could achieve the effect like in [this](https://www.nginx.com/blog/socket-sharding-nginx-release-1-9-1/) Nginx's blog post and pr above
Maybe usage of so_reuseport will low latency of concurrent queries

**Reply by 0xERR0R** [reply]
> 
> I'm all for having a setting to set the number of goroutines, with a dynamic default.

Currently, each incoming request spawns a new goroutine. So basically, there is no limit. I'm thinking about introducing a limit of concurrent requests. For systems under load it could be better to limit a number of concurrent goroutines (incoming requests) to avoid starvations.

**Reply by sethkor** [reply]
Just reading this thread, I'm wondering what the optimal number of requests per blocky instance would be.

I'm running blocky on a 16 core CPU sending it 500QPS and it seems to be consuming 10% CPU which seems like a lot for such a light load
